File: src.java.vadim.ru.leetcode/newStart/leetcode75/CanPlaceFlowers.java
package newStart.leetcode75;
//605. Can Place Flowers
/*
### Алгоритм решения:
1. **Итерируемся по клумбе (`flowerbed`).**
2. **Проверяем возможность посадки цветка в текущую ячейку:**
    - Левая соседняя ячейка (`flowerbed[i - 1]`) пуста или текущая ячейка на самой границе массива.
    - Правая соседняя ячейка (`flowerbed[i + 1]`) пуста или текущая ячейка на противоположной границе массива.

3. Если цветок можно посадить:
    - Сажаем цветок и уменьшаем счётчик `n`.
    - Пропускаем следующую ячейку (так как цветы не могут быть соседними).

4. Если на момент итерации `n` становится равно 0, возвращаем `true` (все цветы посажены).
5. После завершения проверки возвращаем `false`, если `n > 0`.

 */
public class CanPlaceFlowers {
    public static void main(String[] args) {
//        int[] flowerbed = {1,0,0,0,0,0,1};
        int[] flowerbed2 = {1,0,0,0,0,1};
        int n = 2;
        CanPlaceFlowers canPlaceFlowers = new CanPlaceFlowers();
        boolean result = canPlaceFlowers.canPlaceFlowers(flowerbed2, n);
        System.out.println(result);
    }
    public boolean canPlaceFlowers(int[] flowerbed, int n) {
        if (n == 0) return true; // Если не нужно сажать цветы, сразу true

        for (int i = 0; i < flowerbed.length; i++) {
            if (flowerbed[i] == 0) {
                boolean leftEmpty = (i == 0) || (flowerbed[i - 1] == 0);
                boolean rightEmpty = (i == flowerbed.length - 1) || (flowerbed[i + 1] == 0);

                if (leftEmpty && rightEmpty) {
                    flowerbed[i] = 1; // Садим цветок
                    n--; // Уменьшаем количество нужных цветов

                    if (n == 0) return true; // Все цветы посажены

                    i++; // 🔥 Пропускаем следующую ячейку, т.к. уже посадили рядом
                }
            }
        }
        return n == 0; // Если посадили все цветы, true, иначе false
    }
}

File: src.java.vadim.ru.leetcode/newStart/leetcode75/CanPlaceFlowers.py
# 605. Can Place Flowers
class CanPlaceFlowers:
    def can_place_flowers(self, flowerbed, n):
        if n == 0:
            return True  # Если не нужно сажать цветы, сразу true

        i = 0
        while i < len(flowerbed):
            if flowerbed[i] == 0:
                left_empty = (i == 0) or (flowerbed[i - 1] == 0)
                right_empty = (i == len(flowerbed) - 1) or (flowerbed[i + 1] == 0)

                if left_empty and right_empty:
                    flowerbed[i] = 1  # Садим цветок
                    n -= 1  # Уменьшаем количество нужных цветов

                    if n == 0:
                        return True  # Все цветы посажены

                    i += 1  # 🔥 Пропускаем следующую ячейку, т.к. уже посадили рядом
            i += 1
        return n == 0  # Если посадили все цветы, true, иначе false